source/common/http/conn_pool_grid.h

```diff
@@ -24,45 +24,63 @@ class ConnectivityGrid : public ConnectionPool::Instance {
   // auto-retry pools in the case of connection failure.
   //
   // It also relays cancellation calls between the original caller and the
-  // pools currently trying to connect.
-  //
-  // TODO(#15649) currently this tries one connection at a time.
-  // It should instead have a timer of its own and start the second connection
-  // in parallel after a suitable delay.
-  class WrapperCallbacks : public ConnectionPool::Callbacks,
-                           public ConnectionPool::Cancellable,
+  // current connection attempts.
+  class WrapperCallbacks : public ConnectionPool::Cancellable,
                            public LinkedObject<WrapperCallbacks> {
   public:
     WrapperCallbacks(ConnectivityGrid& grid, Http::ResponseDecoder& decoder, PoolIterator pool_it,
                      ConnectionPool::Callbacks& callbacks);
 
-    // ConnectionPool::Callbacks
-    void onPoolFailure(ConnectionPool::PoolFailureReason reason,
-                       absl::string_view transport_failure_reason,
-                       Upstream::HostDescriptionConstSharedPtr host) override;
-    void onPoolReady(RequestEncoder& encoder, Upstream::HostDescriptionConstSharedPtr host,
-                     const StreamInfo::StreamInfo& info,
-                     absl::optional<Http::Protocol> protocol) override;
+    // This holds state for a single connection attempt to a specific pool.
+    struct ConnectionAttemptState : public ConnectionPool::Callbacks,
```

Comment:
It feels weird to me that there are callbacks in the class with "State" in its name as opposed to the one with "Callbacks" in its name. It's also odd that this is a struct not a class. Why the move?